engine/crates/gateway-v2/auth/src/jwt.rs

```diff
@@ -82,93 +120,122 @@ impl Authorizer for JwtProvider {
 
 impl JwtProvider {
     async fn get_access_token(&self, headers: &http::HeaderMap) -> Option<AccessToken> {
+        use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine as _};
+
         let token_str = headers
             .get(&self.config.header_name)
             .and_then(|value| value.to_str().ok())
             .and_then(|value| value.strip_prefix(&self.config.header_value_prefix))?;
 
-        let jose_header = jsonwebtoken::decode_header(token_str).ok()?;
-        let metadata = self.load_metadata().await.ok()?;
-
-        let validation = {
-            let mut validation = jsonwebtoken::Validation::new(jose_header.alg);
-            if let Some(issuer) = self.config.jwks.issuer.as_ref() {
-                validation.set_issuer(&[issuer]);
+        for part in token_str.rsplit('.') {
+            if let Some(v) = URL_SAFE_NO_PAD
+                .decode(part)
+                .ok()
+                .and_then(|s| serde_json::from_slice::<serde_json::Value>(&s).ok())
+            {
+                tracing::debug!("{}", serde_json::to_string_pretty(&v).unwrap());
             }
-            if let Some(audience) = self.config.jwks.audience.as_ref() {
-                validation.set_audience(&[audience]);
-            } else {
-                validation.validate_aud = false;
+        }
+
+        let jwks_bytes = self.load_metadata().await?;
+        let jwks: Jwks<'_> = serde_json::from_slice(&jwks_bytes)
+            .inspect_err(|err| {
+                tracing::debug!("Could not deserialize JWKS: {err}");
+            })
+            .ok()?;
+        let token = decode_token(jwks.keys, UntrustedToken::new(token_str).ok()?)?;
+
+        if let Some(expected) = self.config.jwks.issuer.as_ref() {
+            if token.claims().custom.issuer.as_ref() != Some(expected) {
+                return None;
             }
-            validation.validate_nbf = true;
-            validation
-        };
+        }
 
-        for key in decoding_keys(&metadata.jwks, &jose_header) {
-            if let Ok(TokenData { claims, .. }) = jsonwebtoken::decode(token_str, &key, &validation) {
-                use base64::{engine::general_purpose::URL_SAFE_NO_PAD, Engine as _};
-                let signature = URL_SAFE_NO_PAD
-                    .decode(token_str.rsplit('.').next().expect("valid jwt"))
-                    .expect("valid jwt");
-                return Some(AccessToken::Jwt(JwtToken { claims, signature }));
+        if let Some(expected) = self.config.jwks.audience.as_ref() {
+            let audience = token.claims().custom.audience.as_ref()?;
+            if audience.iter().all(|aud| aud != expected) {
```

Comment:
`all` on an empty iterator returns true - I don't think that's what you want here?

Reply:
damn, good catch, just at the moment I'm merging xD
thanks!

Reply:
ah no that's what I want. I do test it. If there is nothing, it's true and I'm returning `None`

Reply:
not the clearest, have to admit it

Reply:
Oh shit yes, cool.  Ignore me then.